### ISSUE TITLE
Inconsistent master dataset handling

### ISSUE THREAD

**Issue by NathanWang7** [SJTU-DMTai/MASTER#22]
![1111](https://github.com/user-attachments/assets/54f26a40-ecf1-4345-b3ce-03d86bbe45d6)
According to the description in the picture, the processors dealing label only work on learn part (train and valid), which means the label of test data will be original.
But after I loaded dl_test from this [link](https://onedrive.live.com/?id=652674690CC447E6%21sfec60aefc5314c31b4342a4d6d085345&cid=652674690cc447e6&ithint=folder&redeem=aHR0cHM6Ly8xZHJ2Lm1zL2YvYy82NTI2NzQ2OTBjYzQ0N2U2L0V1OEt4djR4eFRGTXREUXFUVzBJVTBVQjhybnBqQUNBNXR3TWk4QkFfUGZiU0E%5FZT1vb2MwemE), it turns out that the label of dl_test is not the same as the original one.
Codes used:
```
universe = 'csi300'
dl_test = pd.read_pickle(f"./data/{universe}/{universe}_dl_test.pkl")
dl_test_df = dl_test.data
dl_test_df["label"].xs(level="instrument", key="SH600000").tail(20)
```
Data got:
	Ref($close, -5) / Ref($close, -1) - 1
datetime	
2022-12-05	-0.140333
2022-12-06	0.203302
2022-12-07	-0.111853
2022-12-08	0.104805
2022-12-09	0.713358
2022-12-12	0.483253
2022-12-13	0.584531
2022-12-14	1.410674
2022-12-15	0.947140
2022-12-16	0.445388
2022-12-19	-0.281419
2022-12-20	-0.413782
2022-12-21	-0.106504
2022-12-22	-0.081041
2022-12-23	0.699521
2022-12-26	0.029281
2022-12-27	-0.228160
2022-12-28	-0.392920
2022-12-29	-0.542747
2022-12-30	-0.307827

Another original stock data downloaded from [repo](https://github.com/chenditc/investment_data/releases) and comfirmed the same with commercial data sources such as ifind or wind. 
Codes here:
```
from qlib.data.dataset.loader import QlibDataLoader
qlib.init()
data_config = {
    "feature": (["$close"], ["$close"]),
    "label": (["Ref($close, -5)/Ref($close, -1) - 1"], ["label"]),
}
data_loader = QlibDataLoader(config=data_config)
df = data_loader.load("csi300", "2022-12-05", "2022-12-30")
df.xs(level="instrument", key="SH600000").tail(20)
```
Data got:
                       feature           label
                       $close            label
datetime		
2022-12-05	10.206157	-0.013618
2022-12-06	10.206157	-0.002658
2022-12-07	10.178172	-0.008112
2022-12-08	10.234141	-0.010935
2022-12-09	10.151119	0.012417
2022-12-12	10.067164	-0.010935
2022-12-13	10.151119	-0.024628
2022-12-14	10.151119	-0.006968
2022-12-15	10.040112	-0.014919
2022-12-16	10.192164	-0.006968
2022-12-19	10.040112	-0.005559
2022-12-20	9.901119	0.004210
2022-12-21	9.970149	0.008269
2022-12-22	10.040112	0.007017
2022-12-23	9.970149	0.025296
2022-12-26	9.846082	0.001397
2022-12-27	10.012127	0.001382
2022-12-28	10.123135	0.015145
2022-12-29	10.040112	0.008224
2022-12-30	10.095149	0.015166

The labels of two data are diferent at all.
So I think maybe the test data from authors is also processed by DropnaLabel, DropExtremeLabel and CSZscoreNorm?

**Comment by NathanWang7** [comment]
Another evidence is stock number per day, csi300 usally contains 300 stocks, the test data only has 284 or 285 stocks, which is just 95% of 300.

**Comment by NathanWang7** [comment]
> we filtered out 5% of most extreme labels in training.
so this describe about data process is not correct.

**Comment by LITONG99** [comment]
Thank you for the report. 

It is of great help. Recently I also received a report that directly using the data in this repo does not produce results exactly matching that in the paper. With you pointing out the published test data only contains around 285 stocks per day, I think I can identify the reason. 

The published validation and test data in this repo are problematic. As I clearly recall, during our experiment, we predicted for around 300 stocks and computed the metrics on every prediction date in the test timespan. I think when I tried to dump the processed data in order to eliminate the reliance on Qlib, I incorrectly used the same processor for the published 'dl_train', 'dl_valid', and 'dl_test'. 

In our original experiment, we used CSZScoreNorm on the test data but did not drop NA or extreme labels. So, the authentic labels also will not be "raw" ( or, I guess "original" in your description). In a realistic environment, one cannot know which labels are extreme or NA in advance. 

In response to the erroneous 'dl_valid' and 'dl_test', I suggest using the Qlib data as I no longer own access to the data source to dump the correct data again. Unfortunately, the efforts I made to refrain the relying on Qlib were not successful due to the mistakes.

Later, I will reorganize this repo to send a notice.

**Comment by NathanWang7** [comment]
Thank you for update the code,there is a problem about your dl_valid:

- The valid data `dl_valid` is generated by the code `dataset.prepare("valid", col_set=["feature", "label], data_key=DataHandlerLP.DK_L)`, which means that it will be processed by `DropnaLabel` as the same as `dl_train`;
- On the other hand, during the model training, `dl_valid` is used by `self.predict()`, which means that it will not be processed by `DropExtremeLabel` and `CSZscoreNorm`, as the same as `dl_test`;
- The evidence is that the length of valid data is 17647 (na labels dropped) instead of 17700 (na labels kept)
- In my option, the valid data shouldn't be processed either by `DropnaLabel` nor `DropExtremeLabel`. Please keep the two processors working or not working together.

**Comment by NathanWang7** [comment]
> Hi, [@NathanWang7](https://github.com/NathanWang7) I just saw your comments.
> 
> I assume that you were mainly suggesting we stop applying the DropnaLabel for the validation set.
> 
> We process the validation set the same as the test set, to estimate the performance of the model on the test set during the validation. Please let me know if my rely makes any difference to your opinion.

Yes, because in real-world scenarios, we will not know which labels were missing during the prediction phase.

**Comment by NathanWang7** [comment]
> Hi, [@NathanWang7](https://github.com/NathanWang7) I think I must be lost in your detailed account to misinterpret it.
> 
> It has been quite a while since I investigated this work. We never claimed anywhere to drop NA labels during prediction. Since MASTER leverages inter-stock correlations, it is **more important** than those methods, which may treat each stock individually, that we did not drop some stocks according to future information. And, the validation and test sets should be processed in the same way.
> 
> Then, I assume you were only referring that the dumped validation data is problematic.
> 
> You are right about `dataset.prepare("valid", col_set=["feature", "label"], data_key=DataHandlerLP.DK_I).` As a quick update, I just checked the Qilb repo, and it seems they are using DataHandlerLP.**DK_L**. I copied their repo to dump the data, so let me remedy the `dl_valid`. It will make no difference to the published models or results, given that we used training loss to stop training in this repo.
> 
> Please let me know if we are on the same page.

Yes, that's the point, thanks for your remedy.
